File: 2-A-Naive/sgemm.c
#define OFF(row, col) row *n + col
const char* sgemm_desc = "A naive C version";

/* This routine performs a dgemm operation
 *  C := C + A * B
 * where A, B, and C are matrices stored in row-major format.
 * On exit, A and B maintain their input values. */    
void square_sgemm (int n, float* A, float* B, float* C)
{
  /* For each row i of C */
  for (int i = 0; i < n; ++i)
    /* For each column j of C */
    for (int j = 0; j < n; ++j) 
     {
     for( int k = 0; k < n; k++ )
	     C[OFF(i,j)] += A[OFF(i,k)] * B[OFF(k,j)]; /* C[i][j] += A[i][k]*B[k][j] */

     }
 }

File: 1-Python/sgemm.py
import numpy as np

# Description of the implementation
sgemm_desc = "A naive Python version"

# Function to perform square matrix multiplication using NumPy
def square_sgemm(A: np.ndarray, B: np.ndarray, C: np.ndarray) -> None:
    """
    This routine performs a sgemm operation using NumPy arrays
    C := C + A * B
    where A, B, and C are square matrices.
    
    Args:
        A: Matrix A as a 2D NumPy array
        B: Matrix B as a 2D NumPy array
        C: Matrix C as a 2D NumPy array (modified in-place)
    """
    n = A.shape[0]

    # Naive implementation (triple loop)
    for i in range(n):
        for j in range(n):
            for k in range(n):
                C[i, j] += A[i, k] * B[k, j]